macros/string_based.go

```diff
@@ -0,0 +1,87 @@
+package macros
+
+import (
+	"strings"
+	"sync"
+)
+
+const (
+	delimiter = "##"
+)
+
+type stringBasedProcessor struct {
+	templates map[string]urlMetaTemplate
+	sync.RWMutex
+}
+
+type urlMetaTemplate struct {
+	indices     []int
+	macroLength []int
+}
+
+func constructTemplate(url string, delimiter string) urlMetaTemplate {
+	currentIndex := 0
+	tmplt := urlMetaTemplate{
+		indices:     []int{},
+		macroLength: []int{},
+	}
+	delimiterLen := len(delimiter)
+	for {
+		currentIndex = currentIndex + strings.Index(url[currentIndex:], delimiter)
+		if currentIndex == -1 {
+			break
+		}
+		middleIndex := currentIndex + delimiterLen
+		endingIndex := strings.Index(url[middleIndex:], delimiter) // ending Delimiter
+		if endingIndex == -1 {
+			break
+		}
+		endingIndex = endingIndex + middleIndex // offset adjustment (Delimiter inclusive)
+		macroLength := endingIndex              // just for readiability
+		tmplt.indices = append(tmplt.indices, currentIndex)
+		tmplt.macroLength = append(tmplt.macroLength, macroLength)
+		currentIndex = endingIndex + 1
+		if currentIndex >= len(url) {
+			break
+		}
+	}
+	return tmplt
+}
+
+func (processor *stringBasedProcessor) Replace(url string, macroProvider *macroProvider) (string, error) {
+	tmplt := processor.getTemplate(url)
+
+	var result strings.Builder
+	currentIndex := 0
+	delimLen := len(delimiter)
+	for i, index := range tmplt.indices {
+		macro := url[index+delimLen : tmplt.macroLength[i]]
+		// copy prev part
+		result.WriteString(url[currentIndex:index])
+		value := macroProvider.GetMacro(macro)
+		if value != "" {
+			result.WriteString(value)
+		}
+		currentIndex = index + len(macro) + 2*delimLen
+	}
+	result.WriteString(url[currentIndex:])
+	return result.String(), nil
+}
+
+func (processor *stringBasedProcessor) getTemplate(url string) urlMetaTemplate {
+	var (
+		template urlMetaTemplate
+		ok       bool
+	)
+	processor.RLock()
+	template, ok = processor.templates[url]
+	processor.RUnlock()
+
+	if !ok {
+		processor.Lock()
+		template = constructTemplate(url, delimiter)
+		processor.templates[url] = template
+		processor.Unlock()
+	}
+	return template
+}
```

Comment:
Why is this a thread safe singleton? URL template should be configured at the host level, meaning it will be the same for all requests, correct? Do you think it makes sense to load it once at the server startup (where you call `NewReplacer`) and jut read it from a config variable?

In case template will come from request, I don't think it's a good idea to keep them all in memory.

 Do I miss something?

Reply:
@VeronikaSolovei9 AS per the PRD, events can be present at host level as well account level.  So processing url template on start up will miss the account level  url templates.
PRD: https://docs.google.com/document/d/1a-U1DytiCTtacM27qH3OZwhPa7bDwQzfpVRHDhyikWw/edit#